Giving season hearts header

oetherington opened the pull request:
Adds the interactive "hearts" header for the end of giving season (desktop only).

The links for the buttons in the header should probably be updated once we know the link for the election results.

<img width="1015" alt="Screenshot 2023-12-23 at 21 56 39" src="https://github.com/ForumMagnum/ForumMagnum/assets/5075628/af90dc92-8b9c-41fe-9450-ca958e41ff1e">
<img width="1015" alt="Screenshot 2023-12-23 at 21 56 43" src="https://github.com/ForumMagnum/ForumMagnum/assets/5075628/de9075ff-5398-487e-bfb1-923c8ea27504">
<img width="228" alt="Screenshot 2023-12-23 at 21 56 48" src="https://github.com/ForumMagnum/ForumMagnum/assets/5075628/ad21277a-687f-4013-b8a5-02bbb3bddf04">
<img width="1019" alt="Screenshot 2023-12-23 at 21 56 28" src="https://github.com/ForumMagnum/ForumMagnum/assets/5075628/9af891a8-7f59-4aec-970d-dac22146ba93">

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1206233966819348) by [Unito](https://www.unito.io)

s-cheng commented:
@oetherington A few notes before I look at the code:

<img width="1438" alt="Screen Shot 2023-12-26 at 3 06 55 PM" src="https://github.com/ForumMagnum/ForumMagnum/assets/9057804/b3f7a68a-ba23-4ad2-9cef-dfabfc131908">

1. Seems like this is appearing on the Giving Portal page, when I assume it should only be on the home page? I think related is a bug where the hearts don't always appear when you go to the home page (like if you navigate from the Giving Portal).
2. It doesn't seem like there's a way for me to remove my heart after I've added it? That seems important to include
3. This header is quite light in dark mode. I know Agnes is out, so I'll just leave it up to you to decide how to handle this. Personally I would make a dark mode version that has a dark red background, but not sure if it's worth the time.

<img width="1064" alt="Screen Shot 2023-12-26 at 3 13 05 PM" src="https://github.com/ForumMagnum/ForumMagnum/assets/9057804/70653cea-c016-4bc0-a2ce-b2061c4b5fbd">

s-cheng commented on packages/lesswrong/components/ea-forum/giving-portal/hooks.ts:
```diff
@@ -83,9 +83,8 @@ export const useAmountRaised = (electionName: string) => {
-  return isEAForum &&
-    moment.utc(timelineSpec.start).isBefore(now) &&
-    moment.utc(timelineSpec.end).isAfter(now);
+  // TODO: Manually disable this after giving season
+  return isEAForum && moment.utc(timelineSpec.start).isBefore(now);
```
Why manually?

s-cheng commented on packages/lesswrong/components/ea-forum/giving-portal/GivingSeasonHeader.tsx:
```diff
@@ -205,7 +241,60 @@ const styles = (theme: ThemeType) => ({
+    cursor: "wait !important",
+  },
+  gsRightHeaderItems: {
+    zIndex: 3, // Appear on top of any hearts
```
This doesn't quite work, ex. this heart prevents me from clicking on search. Looks like you can just decrease `gsHeart`'s `zIndex`.

<img width="1275" alt="Screen Shot 2023-12-26 at 3 25 09 PM" src="https://github.com/ForumMagnum/ForumMagnum/assets/9057804/0df0d979-eb61-4d5c-bcd1-8d76fdf20a5d">

s-cheng commented on packages/lesswrong/components/ea-forum/giving-portal/GivingSeasonHeader.tsx:
```diff
@@ -278,7 +429,76 @@ const GivingSeasonHeader = ({
+    return null;
+  }, [headerRef]);
+
+  const canAddHeart = !!currentUser && !!electionVote?.submittedAt && !isAddingHeart;
```
Do we really want to limit this to only those who voted in the election? I feel like we should allow anyone to add a heart

s-cheng commented on packages/lesswrong/components/ea-forum/giving-portal/GivingSeasonHeader.tsx:
```diff
@@ -371,8 +607,23 @@ const GivingSeasonHeader = ({
+              <div>
+                <span>Add a heart if you got your 2023 donations in</span>
+              </div>
+              <div className={classes.gsButtons}>
+                <EAButton href="/giving-portal">
+                  Donate to the Donation Election winners
+                </EAButton>
+                <EAButton href="/giving-portal#opportunities">
+                  Explore charities
+                </EAButton>
+              </div>
```
FYI [Lizka's post](https://forum.effectivealtruism.org/posts/7D83kwkyaHLQSo6JT/winners-in-the-forum-s-donation-election-2023) about the election winners is out, in case you wanted to link to that here

s-cheng commented on packages/lesswrong/server/repos/DatabaseMetadataRepo.ts:
```diff
@@ -55,4 +56,47 @@ export default class DatabaseMetadataRepo extends AbstractRepo<"DatabaseMetadata
+      randomId(),
+      metadataName,
+      {[userId]: {x, y, theta}},
+    ]);
```
Interesting - seems fine but please have a plan to clean up the code and the db afterwards, I would prefer we don't leave things in random places

s-cheng reviewed (approved):
ASTLAC. I think the only blockers are:
1. Only show this header on the home page
2. Let users remove their heart
3. Make sure the hearts don't interfere with normal header functionality (ex. search)

oetherington commented:
@s-cheng I've just pushed some commits to address your comments.

I'm still having a little trouble with the z-indexes as just lowering the value breaks the tooltips in some places. I'll take a look at it this afternoon before deploying, though it might not be worth blocking deploying depending on how difficult it is to fix.

Thanks!

oetherington commented:
@s-cheng I've just pushed the changes as discussed with Lizka on Slack [here](https://cea-core.slack.com/archives/C03G0AQ925P/p1703680043939129). You may want to give this a last sanity-check if you have time. In the absence of any issues, I'll plan to deploy around 10pm UK time tonight (by which time Lizka will probably be back online to give it a final look over). It's also live on the beta site if you want to try it out there. Thanks again!

<img width="1355" alt="Screenshot 2023-12-27 at 16 50 41" src="https://github.com/ForumMagnum/ForumMagnum/assets/5075628/08781f92-8a3d-4a09-bdaa-eb7dfa9d127d">
<img width="1354" alt="Screenshot 2023-12-27 at 16 50 33" src="https://github.com/ForumMagnum/ForumMagnum/assets/5075628/8a751ae3-d195-437a-92dd-f28d5915fa2d">

s-cheng commented on packages/lesswrong/components/ea-forum/EAHomeRightHandSide.tsx:
```diff
@@ -596,45 +538,11 @@ export const EAHomeRightHandSide = ({classes}: {
   return <AnalyticsContext pageSectionContext="homeRhs">
     {!!currentUser && sidebarToggleNode}
     <div className={classes.root}>
-      {/* TODO: Remove after giving season ends */}
```
Nice - I would also remove the link to the Giving Portal from the mobile menu (though nbd we leave that for a later PR)

s-cheng commented on packages/lesswrong/components/ea-forum/giving-portal/GivingSeasonHeader.tsx:
```diff
@@ -371,8 +638,55 @@ const GivingSeasonHeader = ({
+                  <span className={classes.gsButtonIcon}>{lightbulbIcon}</span>{" "}
+                  EA Infrastructure Fund
+                </EAButton>
+                <EAButton
+                  href="/posts/bBm64htDSKn3ZKiQ5/meet-the-candidates-in-the-forum-s-donation-election-2023"
+                  className={classes.gsSecondaryButton}
+                >
+                  -&gt; Explore more
+                </EAButton>
```
I think this button stands out too much - I suggest just making it a "text button", in the same style as the "Sign up" button at the top right, and remove the arrow in it:

<img width="1437" alt="Screen Shot 2023-12-27 at 12 44 57 PM" src="https://github.com/ForumMagnum/ForumMagnum/assets/9057804/bef1cf15-92ae-4e35-aa09-70795e788ce4">

s-cheng commented on packages/lesswrong/components/ea-forum/giving-portal/GivingSeasonHeader.tsx:
```diff
@@ -371,8 +638,55 @@ const GivingSeasonHeader = ({
+                }
+              </div>
+              <div className={classes.gsButtons}>
+                <EAButton href="https://funds.effectivealtruism.org/funds/animal-welfare">
+                  <span className={classes.gsButtonIcon}>{lightbulbIcon}</span>{" "}
+                  Animal Welfare Fund
+                </EAButton>
+                <EAButton href="https://funds.effectivealtruism.org/funds/far-future">
+                  <span className={classes.gsButtonIcon}>{lightbulbIcon}</span>{" "}
+                  Long-term Future Fund
+                </EAButton>
+                <EAButton href="https://funds.effectivealtruism.org/funds/global-development">
+                  <span className={classes.gsButtonIcon}>{lightbulbIcon}</span>{" "}
+                  Global Development Fund
+                </EAButton>
+                <EAButton href="https://funds.effectivealtruism.org/funds/ea-community">
+                  <span className={classes.gsButtonIcon}>{lightbulbIcon}</span>{" "}
+                  EA Infrastructure Fund
+                </EAButton>
```
I'd also suggest making these icons the same color as the text (dark red) and maybe a bit smaller, and also adding a bit more space above this row of buttons

s-cheng reviewed (approved):
LGTM! I would at least change the "Explore more" button, the rest is less important. I just tested the functionality on the beta site and seems good 👍

Xodarap commented:
This is cool, thanks for doing it

oetherington commented:
@s-cheng Thanks! I addressed the comments and I'm going to go ahead and merge now.

I made the "explore more" button text as you suggested, but I left the arrow as it looked a bit bare without it (I fixed the broken ligature in chrome though). We can tweak further in separate PRs if needed.
